Clarity icons

Wieke opened the pull request:
<!-- Thank you for opening up this pull request! Please make sure to fill out as
much information as you can below.

But, most importantly, please make sure you can say "I did so!" to the
following points:

  - [x] I documented all behaviour as far as I could do.
  - [x] I target the develop-branch, and *not* the master branch.
  - [x] I have tested this extensively and paid extra attention to
        potential cross-platform issues (e.g. Cmd/Ctrl/Super-key bindings)
  - [x] I have made use of ESLint using the provided configuration from the
        repository's .eslintrc.json file, and it did not complain.
  - [x] I have added an entry to the CHANGELOG.md.
  - [x] I matched my code-style to the repository (as far as possible).
  - [x] I do agree that my code will be published under the GNU GPL v3 license.
  - [x] As far as JS-files are concerned, I made sure to copy (in case of new files)
        or adapt (in case of existing files) the info in the header.
  - [x] I synced the latest commits to develop shortly before proposing
        so that no merge issues occur.

  N.B.: Of course you can open a Pull Repository and ask for certain things
  such as file structure later on! It does not need to be perfect on the first
  try :)
 -->

<!-- Below, please shortly describe what the PR does in one or two short sentences. -->
## Description
Adoption of the clarity icon set. Changed existing icons to icons from this set where possible. See issue #719.

<!-- What changes did you make? Please explicitly state any breaking API changes so that nobody is confused why other components suddenly stop working -->
## Changes

Replaced these icons:

* All buttons in the toolbar except pomodoro timer button.
* All buttons, except headers, footnote and remove footnote, in the formatting menu.
* Open directory button in attachments panel.
* Request file, reset field, remove autocorrect, add autocorrect buttons in pdf preferences, project preferences and preferences screens.
* Add button in tags screen.
* Project icon and open/collapse carret in file tree.
* Music note in pomodor timer panel.
* Close button of tabs.

Added a icons utility that is called by the renderer, it loads the clarity module and adds custom icons. Various places where html was being created in javascript where changed (for toolbar, attachment, autocorrect, tabs, filetree icons).

Oh I also added a little hover effect to the open directory button in the attachment pane, so it's a bit more consistent with the other buttons in the application.

<!-- If there is anything else that might be of interest, please provide it here -->
## Additional information

I tested by building and running the app-image. I don't have the means of testing it on windows or mac platforms. I wasn't sure where to place the call to `loadCustomIcons` I ended up putting it at the end of the renderer constructor.

<!-- Please provide any testing system -->
Tested on: Archlinux

nathanlesage commented on source/renderer/util/icons.js:
```diff
@@ -0,0 +1,45 @@
+const path = require('path')
+
+function loadCustomIcons (dir) {
+  fs.readdir(dir, { withFileTypes: true }, (err, list) => {
```
I'd recommend using `fs.promises` here, as it makes the text more legible. And, as we don't have to handle the promises generated by asynchronous functions, this should prove to be no problem (we can catch potential errors and log them, to be sure)

nathanlesage commented on source/renderer/util/icons.js:
```diff
@@ -0,0 +1,45 @@
+  })
+}
+
+module.exports = {
```
This could be simplified by writing

```javascript
module.exports = function () {
    return loadCustomIcons( /* ... */ ) // Return for the promise
}
```

nathanlesage commented on source/renderer/zettlr-renderer.js:
```diff
@@ -79,6 +81,7 @@ class ZettlrRenderer {
         this.beginSearch(term)
       }
     }
+    setTimeout(() => icons.loadCustomIcons(), 0)
```
This can be moved to the init-function where the initial configuration is pulled from main, because I thiiiink that this is only called once … but better check twice …

nathanlesage commented:
I'm a fan! I just cloned it and tested it out locally. I'm really aroused by the looks of the new toolbar, and I could only find super minor things!

![image](https://user-images.githubusercontent.com/17251683/81468794-10e06800-91e2-11ea-818a-371752def790.png)

First, both the file selection icons and the reset-icons are super slightly off to the bottom. Unfortunately, I have no quick fix :/

Then, there are two different clippy icons in the sidebar:

![image](https://user-images.githubusercontent.com/17251683/81468884-72a0d200-91e2-11ea-9ed7-e69d66504ffd.png)

I'm not sure how to solve this (maybe with real file icons based on the attachment's extensions? But Clarity apparently does not include such icons :/)

And last, what about the root file/root directory indicators?

![image](https://user-images.githubusercontent.com/17251683/81468967-e04cfe00-91e2-11ea-8e68-718f9c78510d.png)

Wieke commented:
Yeah I'm not surprised I missed a couple of icons. For example I also forgot the file sort icons:

![image](https://user-images.githubusercontent.com/6737186/81470963-3a06f580-91ee-11ea-8a0e-138002c42d45.png)

Clarity only has a single sort icon though, so I'd probably have to customize something.

![](https://raw.githubusercontent.com/vmware/clarity-assets/master/icons/essential/sort-by-line.svg)

As for the file selection icons, there is some styling on the toolbar buttons that keeps them centred, that might work there as well. Otherwise I suspect fiddling with margins might get it centred.

As for the missed attachment icon, I see a couple of options:

* Reuse the paperclip icon. ![](https://raw.githubusercontent.com/vmware/clarity-assets/master/icons/essential/paperclip-line.svg)
* Just have no icon.
* Use a generic file icon. ![](https://raw.githubusercontent.com/vmware/clarity-assets/master/icons/essential/file-line.svg)
* Use an image icon for images ![](https://raw.githubusercontent.com/vmware/clarity-assets/master/icons/core/image-line.svg), zipped file icon for zipped files ![](https://raw.githubusercontent.com/vmware/clarity-assets/master/icons/essential/file-zip-line.svg) and generic file icons for everything else.

I'd probably go for the generic file icon for the time being. Having icons of specific mimetypes might be useful, though I would probably place that in a new issue cause figuring that out isn't going to be trivial. I did find some nice icons in the [elementary](https://github.com/elementary/icons/tree/master/mimes/symbolic) and [gnome](https://gitlab.gnome.org/GNOME/adwaita-icon-theme/-/tree/master/Adwaita/scalable/mimetypes) themes. (Most of the other filetype icon sets I saw were really colourful and garish, not a good fit for the zettlr style.)

The file/root icons should be easy since there are file and tree-view icons.

![](https://raw.githubusercontent.com/vmware/clarity-assets/master/icons/essential/file-solid.svg) ![](https://raw.githubusercontent.com/vmware/clarity-assets/master/icons/essential/tree-view-line.svg)

I'll probably add the missing icons and process the other feedback within a day or two.

nathanlesage commented:
Perfect, thank you! :)

> Having icons of specific mimetypes might be useful, though I would probably place that in a new issue cause figuring that out isn't going to be trivial.

Concerning this, hot take: How about using the generic file icon and simply add the file extension in there? I mean, it's SVG anyway so we could make use of this. This might actually be pretty helpful as it shows people immediately what file type it is, includes also unknown file extensions and helps if the filename is too long so it's truncated.

Wieke commented:
Oh that's a nice idea I'll give it a try.

Wieke commented:
While fixing the icon not quite being in the center of the button issue I ended up consolidating the `file-select-group` and `form-inline-buttons` classes into a `input-button-group` class, allowing you to do something like this:

```html
<label>Foo</label>
<div class="input-button-group">
  <input type="text">
  <button><clr-icon shape="floppy"></clr-icon></button>
</div>
```
![image](https://user-images.githubusercontent.com/6737186/81505290-6013bf00-92ee-11ea-9e26-0b58d23d4af3.png)

Which I then used to ensure the centeredness of the icons in the buttons that follow text fields. The one drawback though is that this changes the look of the reset and auto correct fields a bit.

Before
![image](https://user-images.githubusercontent.com/6737186/81505535-3c517880-92f0-11ea-991e-3a515ea715f5.png)
After
![image](https://user-images.githubusercontent.com/6737186/81505525-1e841380-92f0-11ea-9147-5a99980f00ed.png)

Before
![image](https://user-images.githubusercontent.com/6737186/81505551-52f7cf80-92f0-11ea-9ef5-0eb611e0d3c6.png)
After
![image](https://user-images.githubusercontent.com/6737186/81505531-25128b00-92f0-11ea-902b-c9271366440c.png)

Now personally I think this consolidation is worth the small change in appearance (which might even be a small improvement, especially in the autocorrect case). But I'd like to get your assent before I commit these changes.

nathanlesage commented:
> I ended up consolidating the file-select-group and form-inline-buttons classes into a input-button-group class

You have my blessings for every simplification of these things! And if they make stuff much more ordered and aligned, you go you! Looks pretty cool tbh.

Wieke commented:
Alright so I made the requested, and some extra, changes.

Loading of custom icons now uses promises.

As described above the `input-button-group` class was added and centeredness of the reset, file request and remove autocorrect buttons was fixed. The centeredness of the add tag/autocorrect buttons was fixed as well. (Presumably, the difference is minimal on my screen.)

The files and directories icons were also replaced. Though I couldn't figure out how to get the root file thing to show up, so no guarantees there (I gave it the same styling as the directories icon, so it shouldn't be too bad).

![image](https://user-images.githubusercontent.com/6737186/81511884-d29b9380-931c-11ea-9f6b-f683886f7ce6.png) (It's a bit small and hard to see on my screen.)

I also reworked the sorting component a bit. The sorting logic was deduplicated by moving it from the file tree/list to the sorting component itself. And the sorting component now emits a `sort-change` event to which the parent listens rather than using `this.$parent` (as far as i know this is a bit more vue-esque).

I also changed the outward behaviour because I thought the current behaviour was a bit vague and I had a hard time figuring out what to replace the icons with. (Honestly I thought the sort-by-time icon meant sort-by-file-size, what with it showing different sized bars.) There are still two toggle buttons/icons but now they ether toggle what is being sorted by (time or name) or what direction the sort goes (ascending or descending). This has the added benefit of the icons showing what way the directory is currently being sorted (something I don't think you could tell before) and it would be slightly easier to add additional sort-by options.

![image](https://user-images.githubusercontent.com/6737186/81512142-8c473400-931e-11ea-82c9-7801843b6d6f.png) (alphabetic ascending)
![image](https://user-images.githubusercontent.com/6737186/81512146-92d5ab80-931e-11ea-9856-520529634dbd.png) (time descending)

The attachment list icon was also replaced. I used your idea of adding the extension to the generic icion. It seems to work pretty well.

![image](https://user-images.githubusercontent.com/6737186/81512222-3aeb7480-931f-11ea-8c40-daf765d4be00.png)

(Only the first 3 letters of the extension are being shown.)

And finally I also replaced the times icons used in the tag cloud dialog and the user dictionary entries list.

Still haven't got them all. The regexes `font-family:.*(font-icon|WebHostingHub-Glyphs)` still has some hits but I'd have to see if I can track them down.

nathanlesage commented:
> There are still two toggle buttons/icons but now they ether toggle what is being sorted by (time or name) or what direction the sort goes (ascending or descending).

This is extremely good. We only need to tell people "Attention, it looks and works slightly different now", albeit I think your solution is even much more straight forward (but you never know :D)

> This has the added benefit of the icons showing what way the directory is currently being sorted (something I don't think you could tell before) and it would be slightly easier to add additional sort-by options.

Yes, agreed!

> (Only the first 3 letters of the extension are being shown.)

This is better than nothing, I think. We MAY want to think about a list of four-or-more-letter-endings that we could shorten down, but that's nothing for right-now!

> The regexes font-family:.*(font-icon|WebHostingHub-Glyphs) still has some hits but I'd have to see if I can track them down.

Yeah, the styling is an utter mess at the moment, we'd have to sort and clean it up at some point …

***

That being set: What about we just merge this today, and have it already in upstream, because then we can already integrate this in the beta phase (might uncover additional issues as well), and you have time to just go and check if you really found all icons!

Wieke commented:
> That being set: What about we just merge this today, and have it already in upstream, because then we can already integrate this in the beta phase (might uncover additional issues as well), and you have time to just go and check if you really found all icons!

Sounds good to me.

nathanlesage commented:
Almost forgot to merge this, but here we go!

Just one small annotation: Feel free to make use of `async`/`await` in the load utility. This'll bring that function more in sync with the rest of the codebase and also reduce that `))))}`-monstrosity at the end of the function :D

Plus we might want to write a test for that function (in case we decide to change something so that we will get a notification if we accidentally break something), that it loads the directory correctly. … or is it failsafe?
